src/materials/GaussianHeatSourceBase.C

```diff
@@ -0,0 +1,258 @@
+  Real dist_z = -2.0 * std::pow(z - z_t, 2.0) / _r[2] / _r[2];
+
+  // Gaussian point heat source
+  return 2.0 * _P * _eta * _f / libMesh::pi / _r[0] / _r[1] / _r[2] *
+         std::exp(dist_x + dist_y + dist_z);
+}
+
+Real
+GaussianHeatSourceBase::computeAveragedHeatSource(
+    const Real x, const Real y, const Real z, const Real time_begin, const Real time_end)
+{
+  mooseAssert(time_end > time_begin, "Begin time should be smaller than end time.");
+
+  unsigned int num_pts = 5;
+  Real Q_integral = 0, Q_integral_old = 0;
+  do
+  {
+    Q_integral_old = Q_integral;
+    Real delta_t = (time_end - time_begin) / num_pts;
+    Real t0 = time_begin;
+    Real Q_begin = computeHeatSourceAtTime(x, y, z, time_begin);
+    Real Q_end;
+    Q_integral = 0;
+    for (unsigned int i = 0; i < num_pts; ++i)
+    {
+      t0 += delta_t;
+      Q_end = computeHeatSourceAtTime(x, y, z, t0);
+      // compute integral of Q between t0 and t0 + delta_t
+      Q_integral += (Q_begin + Q_end) * delta_t / 2.0;
+      // update Q_begin
+      Q_begin = Q_end;
+    }
+    num_pts *= 2;
+    // limit to _number_time_integration pts to accelerate the simulation
+    if (num_pts > _number_time_integration)
+      break;
+  } while (!MooseUtils::absoluteFuzzyEqual(Q_integral_old, Q_integral, 1e-4));
+
+  return Q_integral / (time_end - time_begin);
+}
+
+Real
+GaussianHeatSourceBase::computeMixedHeatSource(
+    const Real x, const Real y, const Real z, const Real time_begin, const Real time_end)
+{
+  mooseAssert(time_end > time_begin, "Begin time should be smaller than end time.");
+
+  // position at time_begin
+  Real x_t0, y_t0, z_t0;
+  computeHeatSourceCenterAtTime(x_t0, y_t0, z_t0, time_begin);
+  Point P_start = Point(x_t0, y_t0, z_t0);
+  // position at time_end
+  Real x_t, y_t, z_t;
+  computeHeatSourceCenterAtTime(x_t, y_t, z_t, time_end);
+  Point P_end = Point(x_t, y_t, z_t);
+
+  Real SE = (P_end - P_start).norm();
+  if (SE < _threshold_length)
+    return computeHeatSourceAtTime(x, y, z, time_end);
+  else
+    return computeAveragedHeatSource(x, y, z, time_begin, time_end);
+}
+
+void
+GaussianHeatSourceBase::computeProperties()
+{
+  // effective radii under current processing parameters
+  if (!_use_input_r)
+  {
+    // compute scanning speed at this time
+    computeHeatSourceMovingSpeedAtTime(_t);
+    // compute the effective radii
+    computeEffectiveRadii(_t);
+  }
+
+  Material::computeProperties();
+}
+
+void
+GaussianHeatSourceBase::computeEffectiveRadii(const Real time)
+{
+  // we do not update _r if we do not proceed in time
+  if (time <= _r_time_prev)
+    return;
+  else
+    _r_time_prev = time;
+
+  // get scaled laser power, scanning speed, and powder feed rate
+  Real lp = _P / 1.0e-3 / 400.0;             // input in unit 1e-3 W
+  Real ss = _scan_speed / 4.23333e-4 / 40.0; // input in mm/ms (1 ipm = 4.23333e-4 mm/ms)
+  Real pf = _feed_rate / 0.031e-3 / 15.0;    // input in g/ms (1rpm = 0.031e-3 g/ms)
```

Comment:
looks like you used a specific unit system. Did you add documents for those and description in the InputParams?